In [5]:
folder_name = 'prev_images'
if not os.path.exists(folder_name):
    os.makedirs(folder_name)

In [6]:
prev_image_url = ['https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv']

PROJETO - DATA WRANGLING: WE RATE DOGS

In [7]:
folder_name = 'prev_images'
if not os.path.exists(folder_name):
    os.makedirs(folder_name)
    
for url in prev_image_url:
    response = requests.get(url)
    with open(os.path.join(folder_name, url.split('/')[-1]), mode = 'wb') as file:
        file.write(response.content)
    
    

In [1]:
import requests
import os
import pandas as pd
import numpy as np
try:
    import tweepy
except:
    import pip
    pip.main(['install','tweepy'])
    import tweepy
from PIL import Image
from io import BytesIO

In [8]:
os.listdir(folder_name)

['image-predictions.tsv']

In [9]:
open ('C:\\Users\\Samsung\\Desktop\\Alexsander\\UDACITY\\Fundamentos de Data Science II\\PROJETOS\\DATA WRANGLING\\prev_images\\image-predictions.tsv','r')

<_io.TextIOWrapper name='C:\\Users\\Samsung\\Desktop\\Alexsander\\UDACITY\\Fundamentos de Data Science II\\PROJETOS\\DATA WRANGLING\\prev_images\\image-predictions.tsv' mode='r' encoding='cp1252'>

In [10]:
images = pd.read_csv('C:\\Users\\Samsung\\Desktop\\Alexsander\\UDACITY\\Fundamentos de Data Science II\\PROJETOS\\DATA WRANGLING\\prev_images\\image-predictions.tsv', sep='\t', encoding = 'utf-8')

In [11]:
images.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [12]:
twitter_links = pd.read_csv('twitter-archive-enhanced.csv', encoding = 'utf-8')
twitter_links

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [13]:
# Consumer keys and access tokens, used for OAuth
consumer_key = 'w8SbEeKSm54fE3G52SIE0eOW6'
consumer_secret = 'GMte3O13adlIGJngWVfATy1MAGMMya5czIoSznRk9AMdPXIJLe'
access_token = '524884564-oi1RpFB38FEozm6Oy0oIPbQyNCEwGxrRCShaXQJF'
access_token_secret = 'DH0qeVFYJ9GtFrrySEjhaLIwHY15Q5rwt9XdFgFSuDLgv'

In [14]:
# OAuth process, using the keys and tokens
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)
api = tweepy.API(auth,  parser = tweepy.parsers.JSONParser(), wait_on_rate_limit = True)

In [15]:
tweets_list = []
errors = []

for tweet_id in twitter_links['tweet_id']:
    try:
        page = api.get_status(tweet_id, tweet_mode = 'extended')
        curtidas = page['favorite_count']
        retweets = page['retweet_count']
        data = page['created_at']
        
        tweets_list.append({'tweet_id': int(tweet_id),
                        'curtidas': int(curtidas),
                        'retweets': int(retweets),
                        'data': pd.to_datetime(data)})
    
    except Exception as e:
        print(str(tweet_id)+ " _ " + str(e))
        errors.append(tweet_id)


888202515573088257 _ [{'code': 144, 'message': 'No status found with that ID.'}]
873697596434513921 _ [{'code': 144, 'message': 'No status found with that ID.'}]
872668790621863937 _ [{'code': 34, 'message': 'Sorry, that page does not exist.'}]
869988702071779329 _ [{'code': 144, 'message': 'No status found with that ID.'}]
866816280283807744 _ [{'code': 144, 'message': 'No status found with that ID.'}]
861769973181624320 _ [{'code': 144, 'message': 'No status found with that ID.'}]
845459076796616705 _ [{'code': 144, 'message': 'No status found with that ID.'}]
842892208864923648 _ [{'code': 144, 'message': 'No status found with that ID.'}]
837012587749474308 _ [{'code': 144, 'message': 'No status found with that ID.'}]
827228250799742977 _ [{'code': 144, 'message': 'No status found with that ID.'}]
802247111496568832 _ [{'code': 144, 'message': 'No status found with that ID.'}]
775096608509886464 _ [{'code': 144, 'message': 'No status found with that ID.'}]
771004394259247104 _ [{'co

In [16]:
print("The lengh of the result", len(tweets_list))
print("The lengh of the errors", len(errors))

The lengh of the result 2341
The lengh of the errors 15


In [22]:
add_error = []
for e in errors:
    try:
        page = api.get_status(tweet_id, tweet_mode = 'extended')
        curtidas = page['favorite_count']
        retweets = page['retweet_count']
        data = page['created_at']
        
        tweets_list.append({'tweet_id': int(tweet_id),
                        'curtidas': int(curtidas),
                        'retweets': int(retweets),
                        'data': pd.to_datetime(data)})
        
    except Exception:
        print(str(tweet_id)+ " _ " + str(e))
        add_error.append(e)

In [23]:
json_tweets = pd.DataFrame(tweets_list, columns = ['tweet_id', 'curtidas', 'retweets', 'data'])
json_tweets.to_csv('tweet_json.txt', encoding = 'utf-8', index=False)

In [24]:
json_tweets = pd.read_csv('tweet_json.txt', encoding = 'utf-8')

In [25]:
json_tweets

tweet_id  curtidas  retweets                 data
0     892420643555336193     38385      8431  2017-08-01 16:23:56
1     892177421306343426     32888      6211  2017-08-01 00:17:27
2     891815181378084864     24784      4111  2017-07-31 00:18:03
3     891689557279858688     41720      8557  2017-07-30 15:58:51
4     891327558926688256     39894      9284  2017-07-29 16:00:24
5     891087950875897856     20022      3081  2017-07-29 00:08:17
6     890971913173991426     11722      2045  2017-07-28 16:27:12
7     890729181411237888     64793     18693  2017-07-28 00:22:40
8     890609185150312448     27533      4223  2017-07-27 16:25:51
9     890240255349198849     31602      7319  2017-07-26 15:59:51
10    890006608113172480     30344      7262  2017-07-26 00:31:25
11    889880896479866881     27498      4927  2017-07-25 16:11:53
12    889665388333682689     47629      9962  2017-07-25 01:55:32
13    889638837579907072     26895      4494  2017-07-25 00:10:02
14    889531135344209921     14942      2214  2017-07-24 17:02:04
15    889278841981685760     25040      5365  2017-07-24 00:19:32
16    888917238123831296     28794      4457  2017-07-23 00:22:39
17    888804989199671297     25333      4279  2017-07-22 16:56:37
18    888554962724278272     19674      3540  2017-07-22 00:23:06
19    888078434458587136     21539      3465  2017-07-20 16:49:33
20    887705289381826560     29856      5333  2017-07-19 16:06:48
21    887517139158093824     45788     11552  2017-07-19 03:39:09
22    887473957103951883     68395     18018  2017-07-19 00:47:34
23    887343217045368832     33334     10299  2017-07-18 16:08:03
24    887101392804085760     30241      5907  2017-07-18 00:07:08
25    886983233522544640     34808      7704  2017-07-17 16:17:36
26    886736880519319552     11932      3258  2017-07-16 23:58:41
27    886680336477933568     22195      4421  2017-07-16 20:14:00
28    886366144734445568     20981      3171  2017-07-15 23:25:31
29    886267009285017600       116         4  2017-07-15 16:51:35
...                  ...       ...       ...                  ...
2341  666020888022790149      2554       508  2015-11-15 22:32:08
2342  666020888022790149      2554       508  2015-11-15 22:32:08
2343  666020888022790149      2554       508  2015-11-15 22:32:08
2344  666020888022790149      2554       508  2015-11-15 22:32:08
2345  666020888022790149      2554       508  2015-11-15 22:32:08
2346  666020888022790149      2554       508  2015-11-15 22:32:08
2347  666020888022790149      2554       508  2015-11-15 22:32:08
2348  666020888022790149      2554       508  2015-11-15 22:32:08
2349  666020888022790149      2554       508  2015-11-15 22:32:08
2350  666020888022790149      2554       508  2015-11-15 22:32:08
2351  666020888022790149      2554       508  2015-11-15 22:32:08
2352  666020888022790149      2554       508  2015-11-15 22:32:08
2353  666020888022790149      2554       508  2015-11-15 22:32:08
2354  666020888022790149      2554       508  2015-11-15 22:32:08
2355  666020888022790149      2554       508  2015-11-15 22:32:08
2356  666020888022790149      2554       508  2015-11-15 22:32:08
2357  666020888022790149      2554       508  2015-11-15 22:32:08
2358  666020888022790149      2554       508  2015-11-15 22:32:08
2359  666020888022790149      2554       508  2015-11-15 22:32:08
2360  666020888022790149      2554       508  2015-11-15 22:32:08
2361  666020888022790149      2554       508  2015-11-15 22:32:08
2362  666020888022790149      2554       508  2015-11-15 22:32:08
2363  666020888022790149      2554       508  2015-11-15 22:32:08
2364  666020888022790149      2554       508  2015-11-15 22:32:08
2365  666020888022790149      2554       508  2015-11-15 22:32:08
2366  666020888022790149      2554       508  2015-11-15 22:32:08
2367  666020888022790149      2554       508  2015-11-15 22:32:08
2368  666020888022790149      2554       508  2015-11-15 22:32:08
2369  666020888022790149      2554       508  2015-11-15

In [26]:
json_tweets

tweet_id  curtidas  retweets                 data
0     892420643555336193     38385      8431  2017-08-01 16:23:56
1     892177421306343426     32888      6211  2017-08-01 00:17:27
2     891815181378084864     24784      4111  2017-07-31 00:18:03
3     891689557279858688     41720      8557  2017-07-30 15:58:51
4     891327558926688256     39894      9284  2017-07-29 16:00:24
5     891087950875897856     20022      3081  2017-07-29 00:08:17
6     890971913173991426     11722      2045  2017-07-28 16:27:12
7     890729181411237888     64793     18693  2017-07-28 00:22:40
8     890609185150312448     27533      4223  2017-07-27 16:25:51
9     890240255349198849     31602      7319  2017-07-26 15:59:51
10    890006608113172480     30344      7262  2017-07-26 00:31:25
11    889880896479866881     27498      4927  2017-07-25 16:11:53
12    889665388333682689     47629      9962  2017-07-25 01:55:32
13    889638837579907072     26895      4494  2017-07-25 00:10:02
14    889531135344209921     14942      2214  2017-07-24 17:02:04
15    889278841981685760     25040      5365  2017-07-24 00:19:32
16    888917238123831296     28794      4457  2017-07-23 00:22:39
17    888804989199671297     25333      4279  2017-07-22 16:56:37
18    888554962724278272     19674      3540  2017-07-22 00:23:06
19    888078434458587136     21539      3465  2017-07-20 16:49:33
20    887705289381826560     29856      5333  2017-07-19 16:06:48
21    887517139158093824     45788     11552  2017-07-19 03:39:09
22    887473957103951883     68395     18018  2017-07-19 00:47:34
23    887343217045368832     33334     10299  2017-07-18 16:08:03
24    887101392804085760     30241      5907  2017-07-18 00:07:08
25    886983233522544640     34808      7704  2017-07-17 16:17:36
26    886736880519319552     11932      3258  2017-07-16 23:58:41
27    886680336477933568     22195      4421  2017-07-16 20:14:00
28    886366144734445568     20981      3171  2017-07-15 23:25:31
29    886267009285017600       116         4  2017-07-15 16:51:35
...                  ...       ...       ...                  ...
2341  666020888022790149      2554       508  2015-11-15 22:32:08
2342  666020888022790149      2554       508  2015-11-15 22:32:08
2343  666020888022790149      2554       508  2015-11-15 22:32:08
2344  666020888022790149      2554       508  2015-11-15 22:32:08
2345  666020888022790149      2554       508  2015-11-15 22:32:08
2346  666020888022790149      2554       508  2015-11-15 22:32:08
2347  666020888022790149      2554       508  2015-11-15 22:32:08
2348  666020888022790149      2554       508  2015-11-15 22:32:08
2349  666020888022790149      2554       508  2015-11-15 22:32:08
2350  666020888022790149      2554       508  2015-11-15 22:32:08
2351  666020888022790149      2554       508  2015-11-15 22:32:08
2352  666020888022790149      2554       508  2015-11-15 22:32:08
2353  666020888022790149      2554       508  2015-11-15 22:32:08
2354  666020888022790149      2554       508  2015-11-15 22:32:08
2355  666020888022790149      2554       508  2015-11-15 22:32:08
2356  666020888022790149      2554       508  2015-11-15 22:32:08
2357  666020888022790149      2554       508  2015-11-15 22:32:08
2358  666020888022790149      2554       508  2015-11-15 22:32:08
2359  666020888022790149      2554       508  2015-11-15 22:32:08
2360  666020888022790149      2554       508  2015-11-15 22:32:08
2361  666020888022790149      2554       508  2015-11-15 22:32:08
2362  666020888022790149      2554       508  2015-11-15 22:32:08
2363  666020888022790149      2554       508  2015-11-15 22:32:08
2364  666020888022790149      2554       508  2015-11-15 22:32:08
2365  666020888022790149      2554       508  2015-11-15 22:32:08
2366  666020888022790149      2554       508  2015-11-15 22:32:08
2367  666020888022790149      2554       508  2015-11-15 22:32:08
2368  666020888022790149      2554       508  2015-11-15 22:32:08
2369  666020888022790149      2554       508  2015-11-15

**ASSESS**

**Após a coleta dos dados, realizamos a avaliação dos mesmos de forma manual e programática** 

1. Verificar dados sujos e bagunçados

In [27]:
images

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [28]:
images.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [29]:
images.nunique()

tweet_id    2075
jpg_url     2009
img_num        4
p1           378
p1_conf     2006
p1_dog         2
p2           405
p2_conf     2004
p2_dog         2
p3           408
p3_conf     2006
p3_dog         2
dtype: int64

In [30]:
images[images.duplicated(['tweet_id'], keep=False)]['tweet_id']

Series([], Name: tweet_id, dtype: int64)

In [31]:
images['jpg_url'].value_counts()

https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg                                            2
https://pbs.twimg.com/media/C12whDoVEAALRxa.jpg                                            2
https://pbs.twimg.com/media/CVgdFjNWEAAxmbq.jpg                                            2
https://pbs.twimg.com/media/Cs_DYr1XEAA54Pu.jpg                                            2
https://pbs.twimg.com/media/CWyD2HGUYAQ1Xa7.jpg                                            2
https://pbs.twimg.com/media/CkjMx99UoAM2B1a.jpg                                            2
https://pbs.twimg.com/media/CcG07BYW0AErrC9.jpg                                            2
https://pbs.twimg.com/media/CUN4Or5UAAAa5K4.jpg                                            2
https://pbs.twimg.com/ext_tw_video_thumb/815965888126062592/pu/img/JleSw4wRhgKDWQj5.jpg    2
https://pbs.twimg.com/media/CsGnz64WYAEIDHJ.jpg                                            2
https://pbs.twimg.com/media/CVuQ2LeUsAAIe3s.jpg                       

In [32]:
images[images.duplicated(['jpg_url'], keep=False)]['jpg_url']

85        https://pbs.twimg.com/media/CUN4Or5UAAAa5K4.jpg
224       https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg
241       https://pbs.twimg.com/media/CU3mITUWIAAfyQS.jpg
327       https://pbs.twimg.com/media/CVMOlMiWwAA4Yxl.jpg
382       https://pbs.twimg.com/media/CVgdFjNWEAAxmbq.jpg
432       https://pbs.twimg.com/media/CVuQ2LeUsAAIe3s.jpg
480     https://pbs.twimg.com/ext_tw_video_thumb/67535...
487       https://pbs.twimg.com/media/CV_cnjHWUAADc-c.jpg
587       https://pbs.twimg.com/media/CWyD2HGUYAQ1Xa7.jpg
591       https://pbs.twimg.com/media/CWza7kpWcAAdYLc.jpg
602       https://pbs.twimg.com/media/CW88XN4WsAAlo8r.jpg
713       https://pbs.twimg.com/media/CYLDikFWEAAIy1y.jpg
800       https://pbs.twimg.com/media/CZhn-QAWwAASQan.jpg
915       https://pbs.twimg.com/media/Cbs3DOAXIAAp3Bd.jpg
930       https://pbs.twimg.com/media/CcG07BYW0AErrC9.jpg
985       https://pbs.twimg.com/media/CdHwZd0VIAA4792.jpg
1033    https://pbs.twimg.com/tweet_video_thumb/CeBym7...
1045      http

In [33]:
twitter_links

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [34]:
twitter_links.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [35]:
twitter_links.nunique()

tweet_id                      2356
in_reply_to_status_id           77
in_reply_to_user_id             31
timestamp                     2356
source                           4
text                          2356
retweeted_status_id            181
retweeted_status_user_id        25
retweeted_status_timestamp     181
expanded_urls                 2218
rating_numerator                40
rating_denominator              18
name                           957
doggo                            2
floofer                          2
pupper                           2
puppo                            2
dtype: int64

In [36]:
twitter_links[twitter_links.duplicated(['expanded_urls'], keep=False)]['expanded_urls']

30                                                    NaN
55                                                    NaN
64                                                    NaN
68      https://twitter.com/dog_rates/status/878057613...
73      https://www.gofundme.com/3yd6y1c,https://twitt...
74      https://twitter.com/dog_rates/status/669000397...
75      https://www.gofundme.com/3yd6y1c,https://twitt...
76      https://twitter.com/dog_rates/status/878057613...
91      https://twitter.com/dog_rates/status/866334964...
97      https://www.gofundme.com/help-my-baby-sierra-g...
98      https://www.gofundme.com/help-my-baby-sierra-g...
109     https://twitter.com/dog_rates/status/841077006...
113                                                   NaN
118     https://twitter.com/dog_rates/status/859196978...
124     https://www.gofundme.com/3ti3nps,https://twitt...
126     https://www.gofundme.com/3ti3nps,https://twitt...
132     https://twitter.com/dog_rates/status/866450705...
135     https:

In [37]:
twitter_links[twitter_links.duplicated(['name'], keep=False)]['name']

0        Phineas
2         Archie
4       Franklin
5           None
6            Jax
7           None
8           Zoey
9         Cassie
10          Koda
12          None
13           Ted
15        Oliver
17          Zeke
19        Canela
20        Gerald
21       Jeffrey
23        Canela
24          None
25          None
28         Derek
30          None
31       Waffles
32          None
35          None
36         Lilly
37          None
38          Earl
39          Lola
40         Kevin
41          None
          ...   
2326       quite
2327           a
2328        None
2329        None
2330        None
2331        None
2332        None
2333          an
2334           a
2335          an
2336        None
2337        None
2338        None
2339        None
2340        None
2341        None
2342        None
2343        None
2344        None
2345         the
2346         the
2347           a
2348           a
2349          an
2350           a
2351        None
2352           a
2353          

In [38]:
twitter_links.loc[(twitter_links['name'].str.islower())]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
22    887517139158093824                    NaN                  NaN   
56    881536004380872706                    NaN                  NaN   
118   869988702071779329                    NaN                  NaN   
169   859196978902773760                    NaN                  NaN   
193   855459453768019968                    NaN                  NaN   
335   832645525019123713                    NaN                  NaN   
369   828650029636317184                    NaN                  NaN   
542   806219024703037440                    NaN                  NaN   
649   792913359805018113                    NaN                  NaN   
682   788552643979468800                    NaN                  NaN   
759   778396591732486144                    NaN                  NaN   
773   776249906839351296                    NaN                  NaN   
801   772581559778025472                    NaN                  NaN   
819   770655142660169732                    NaN                  NaN   
822   770093767776997377                    NaN                  NaN   
852   765395769549590528                    NaN                  NaN   
924   755206590534418437                    NaN                  NaN   
988   748977405889503236                    NaN                  NaN   
992   748692773788876800                    NaN                  NaN   
993   748575535303884801                    NaN                  NaN   
1002  747885874273214464                    NaN                  NaN   
1004  747816857231626240                    NaN                  NaN   
1017  746872823977771008                    NaN                  NaN   
1025  746369468511756288                    NaN                  NaN   
1031  745422732645535745                    NaN                  NaN   
1040  744223424764059648                    NaN                  NaN   
1049  743222593470234624                    NaN                  NaN   
1063  741067306818797568                    NaN                  NaN   
1071  740214038584557568                    NaN                  NaN   
1095  736392552031657984                    NaN                  NaN   
...                  ...                    ...                  ...   
2191  668955713004314625                    NaN                  NaN   
2198  668815180734689280                    NaN                  NaN   
2204  668636665813057536                    NaN                  NaN   
2211  668614819948453888                    NaN                  NaN   
2212  668587383441514497                    NaN                  NaN   
2218  668507509523615744                    NaN                  NaN   
2222  668466899341221888                    NaN                  NaN   
2235  668171859951755264                    NaN                  NaN   
2249  667861340749471744                    NaN                  NaN   
2255  667773195014021121                    NaN                  NaN   
2264  667538891197542400                    NaN                  NaN   
2273  667470559035432960                    NaN                  NaN   
2287  667177989038297088                    NaN                  NaN   
2304  666983947667116034                    NaN                  NaN   
2311  666781792255496192                    NaN                  NaN   
2314  666701168228331520                    NaN                  NaN   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN         

In [39]:
twitter_links.name.sort_values()

1035            Abby
1021            Abby
938              Ace
1933            Acro
1327           Adele
1934           Aiden
77               Aja
480            Akumi
820               Al
875           Albert
1954          Albert
144            Albus
412            Albus
1115         Aldrick
2046       Alejandro
374        Alexander
1189    Alexanderson
486              Alf
2238           Alfie
661            Alfie
367            Alfie
1616           Alfie
858            Alfie
51              Alfy
201            Alice
1701           Alice
1495           Amber
1334         Ambrose
2078             Amy
2195          Amélie
            ...     
852               my
988              not
335              not
1747      officially
1206             old
993              one
1936             one
924              one
369              one
2326           quite
169            quite
118            quite
193            quite
2030           space
22              such
2345             the
2346         

In [40]:
json_tweets

tweet_id  curtidas  retweets                 data
0     892420643555336193     38385      8431  2017-08-01 16:23:56
1     892177421306343426     32888      6211  2017-08-01 00:17:27
2     891815181378084864     24784      4111  2017-07-31 00:18:03
3     891689557279858688     41720      8557  2017-07-30 15:58:51
4     891327558926688256     39894      9284  2017-07-29 16:00:24
5     891087950875897856     20022      3081  2017-07-29 00:08:17
6     890971913173991426     11722      2045  2017-07-28 16:27:12
7     890729181411237888     64793     18693  2017-07-28 00:22:40
8     890609185150312448     27533      4223  2017-07-27 16:25:51
9     890240255349198849     31602      7319  2017-07-26 15:59:51
10    890006608113172480     30344      7262  2017-07-26 00:31:25
11    889880896479866881     27498      4927  2017-07-25 16:11:53
12    889665388333682689     47629      9962  2017-07-25 01:55:32
13    889638837579907072     26895      4494  2017-07-25 00:10:02
14    889531135344209921     14942      2214  2017-07-24 17:02:04
15    889278841981685760     25040      5365  2017-07-24 00:19:32
16    888917238123831296     28794      4457  2017-07-23 00:22:39
17    888804989199671297     25333      4279  2017-07-22 16:56:37
18    888554962724278272     19674      3540  2017-07-22 00:23:06
19    888078434458587136     21539      3465  2017-07-20 16:49:33
20    887705289381826560     29856      5333  2017-07-19 16:06:48
21    887517139158093824     45788     11552  2017-07-19 03:39:09
22    887473957103951883     68395     18018  2017-07-19 00:47:34
23    887343217045368832     33334     10299  2017-07-18 16:08:03
24    887101392804085760     30241      5907  2017-07-18 00:07:08
25    886983233522544640     34808      7704  2017-07-17 16:17:36
26    886736880519319552     11932      3258  2017-07-16 23:58:41
27    886680336477933568     22195      4421  2017-07-16 20:14:00
28    886366144734445568     20981      3171  2017-07-15 23:25:31
29    886267009285017600       116         4  2017-07-15 16:51:35
...                  ...       ...       ...                  ...
2341  666020888022790149      2554       508  2015-11-15 22:32:08
2342  666020888022790149      2554       508  2015-11-15 22:32:08
2343  666020888022790149      2554       508  2015-11-15 22:32:08
2344  666020888022790149      2554       508  2015-11-15 22:32:08
2345  666020888022790149      2554       508  2015-11-15 22:32:08
2346  666020888022790149      2554       508  2015-11-15 22:32:08
2347  666020888022790149      2554       508  2015-11-15 22:32:08
2348  666020888022790149      2554       508  2015-11-15 22:32:08
2349  666020888022790149      2554       508  2015-11-15 22:32:08
2350  666020888022790149      2554       508  2015-11-15 22:32:08
2351  666020888022790149      2554       508  2015-11-15 22:32:08
2352  666020888022790149      2554       508  2015-11-15 22:32:08
2353  666020888022790149      2554       508  2015-11-15 22:32:08
2354  666020888022790149      2554       508  2015-11-15 22:32:08
2355  666020888022790149      2554       508  2015-11-15 22:32:08
2356  666020888022790149      2554       508  2015-11-15 22:32:08
2357  666020888022790149      2554       508  2015-11-15 22:32:08
2358  666020888022790149      2554       508  2015-11-15 22:32:08
2359  666020888022790149      2554       508  2015-11-15 22:32:08
2360  666020888022790149      2554       508  2015-11-15 22:32:08
2361  666020888022790149      2554       508  2015-11-15 22:32:08
2362  666020888022790149      2554       508  2015-11-15 22:32:08
2363  666020888022790149      2554       508  2015-11-15 22:32:08
2364  666020888022790149      2554       508  2015-11-15 22:32:08
2365  666020888022790149      2554       508  2015-11-15 22:32:08
2366  666020888022790149      2554       508  2015-11-15 22:32:08
2367  666020888022790149      2554       508  2015-11-15 22:32:08
2368  666020888022790149      2554       508  2015-11-15 22:32:08
2369  666020888022790149      2554       508  2015-11-15

In [41]:
json_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2371 entries, 0 to 2370
Data columns (total 4 columns):
tweet_id    2371 non-null int64
curtidas    2371 non-null int64
retweets    2371 non-null int64
data        2371 non-null object
dtypes: int64(3), object(1)
memory usage: 74.2+ KB


In [42]:
json_tweets.nunique()

tweet_id    2341
curtidas    2027
retweets    1724
data        2341
dtype: int64

In [43]:
json_tweets[json_tweets.duplicated(['tweet_id'], keep=False)]

tweet_id  curtidas  retweets                 data
2340  666020888022790149      2554       508  2015-11-15 22:32:08
2341  666020888022790149      2554       508  2015-11-15 22:32:08
2342  666020888022790149      2554       508  2015-11-15 22:32:08
2343  666020888022790149      2554       508  2015-11-15 22:32:08
2344  666020888022790149      2554       508  2015-11-15 22:32:08
2345  666020888022790149      2554       508  2015-11-15 22:32:08
2346  666020888022790149      2554       508  2015-11-15 22:32:08
2347  666020888022790149      2554       508  2015-11-15 22:32:08
2348  666020888022790149      2554       508  2015-11-15 22:32:08
2349  666020888022790149      2554       508  2015-11-15 22:32:08
2350  666020888022790149      2554       508  2015-11-15 22:32:08
2351  666020888022790149      2554       508  2015-11-15 22:32:08
2352  666020888022790149      2554       508  2015-11-15 22:32:08
2353  666020888022790149      2554       508  2015-11-15 22:32:08
2354  666020888022790149      2554       508  2015-11-15 22:32:08
2355  666020888022790149      2554       508  2015-11-15 22:32:08
2356  666020888022790149      2554       508  2015-11-15 22:32:08
2357  666020888022790149      2554       508  2015-11-15 22:32:08
2358  666020888022790149      2554       508  2015-11-15 22:32:08
2359  666020888022790149      2554       508  2015-11-15 22:32:08
2360  666020888022790149      2554       508  2015-11-15 22:32:08
2361  666020888022790149      2554       508  2015-11-15 22:32:08
2362  666020888022790149      2554       508  2015-11-15 22:32:08
2363  666020888022790149      2554       508  2015-11-15 22:32:08
2364  666020888022790149      2554       508  2015-11-15 22:32:08
2365  666020888022790149      2554       508  2015-11-15 22:32:08
2366  666020888022790149      2554       508  2015-11-15 22:32:08
2367  666020888022790149      2554       508  2015-11-15 22:32:08
2368  666020888022790149      2554       508  2015-11-15 22:32:08
2369  666020888022790149      2554       508  2015-11-15 22:32:08
2370  666020888022790149      2554       508  2015-11-15 22:32:08

QUALIDADE

1. Coluna criada para doggo, floofer, pupper, puppo contém variáveis diferentes. Após análise, adicionei duas as classificações existentes e deixei uma variável chamada 
2. Eliminei os Retweets.
3. Em imagens, p1, p2 e p3 podem ser strings ao invés de objects.
4. Renomear as colunas de previsão da raça do cão.
5. Coluna nome com campos none e/ou letras ou palavras sem sentido.
6. Eliminar url’s de imagens duplicadas
7. Alterar o tipo de dado dos campos numerador e denominador
8. Alterar valores incorretos nos campos numerador e denominador
9. Criar nova variável Avaliação, que é obtido com o cálculo do numerador pelo denominador.
10. Mudar a variável timestamp para data.



ARRUMAÇÃO

1.As colunas doggo, floofer, pupper, puppo devem ser somente uma variável.

2.Podemos juntar todas as planilhas em uma planilha mestre para facilitar a análise.

3.Retirei as colunas 'retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp','in_reply_to_status_id','in_reply_to_user_id', 'retweeted_status_id'




Vamos copiar os dados, para controle e fácil exportação.

In [44]:
twitter_links_copy = twitter_links.copy()
images_copy = images.copy()
json_tweets_copy = json_tweets.copy()

Limpeza


1. Vamos começar unindo as colunas doggo, floofer, pupper, puppo em uma variavel.
Achei ideal começar com esta coluna, pois assim posso eliminar possiveis campos Null que teriamos nas colunas.

In [45]:
twitter_links_copy.replace('None', np.nan, inplace=True)
twitter_links_copy['dog_classification'] = twitter_links_copy[['doggo', 'floofer', 'pupper', 'puppo']].fillna('').sum(axis=1).astype(str)
twitter_links_copy = twitter_links_copy.drop(['doggo', 'floofer', 'pupper', 'puppo'], axis=1)

In [46]:
twitter_links_copy

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [47]:
twitter_links_copy['dog_classification'].value_counts()

                1976
pupper           245
doggo             83
puppo             29
doggopupper       12
floofer            9
doggofloofer       1
doggopuppo         1
Name: dog_classification, dtype: int64

Em 1976 campos, conseguimos 380 classificações.
Destas 380, temos 12 doggopupper, 1 doggopuppo e 1 doggofloffer.
Vamos verificar estas linhas.

In [51]:
twitter_links_copy[twitter_links_copy['dog_classification'] == 'doggofloofer'] 

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
200  854010172552949760                    NaN                  NaN   

                     timestamp  \
200  2017-04-17 16:34:26 +0000   

                                                source  \
200  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
200  At first I thought this was a shy doggo, but i...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
200                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
200  https://twitter.com/dog_rates/status/854010172...                11   

     rating_denominator name dog_classification  
200                  10  NaN       doggofloofer

In [52]:
twitter_links_copy[twitter_links_copy['dog_classification'] == 'doggopuppo'] 

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
191  855851453814013952                    NaN                  NaN   

                     timestamp  \
191  2017-04-22 18:31:02 +0000   

                                                source  \
191  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
191  Here's a puppo participating in the #ScienceMa...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
191                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
191  https://twitter.com/dog_rates/status/855851453...                13   

     rating_denominator name dog_classification  
191                  10  NaN         doggopuppo

In [53]:
twitter_links_copy[twitter_links_copy['dog_classification'] == 'doggopupper'] 

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
460   817777686764523521                    NaN                  NaN   
531   808106460588765185                    NaN                  NaN   
565   802265048156610565           7.331095e+17         4.196984e+09   
575   801115127852503040                    NaN                  NaN   
705   785639753186217984                    NaN                  NaN   
733   781308096455073793                    NaN                  NaN   
778   775898661951791106                    NaN                  NaN   
822   770093767776997377                    NaN                  NaN   
889   759793422261743616                    NaN                  NaN   
956   751583847268179968                    NaN                  NaN   
1063  741067306818797568                    NaN                  NaN   
1113  733109485275860992                    NaN                  NaN   

                      timestamp  \
460   2017-01-07 16:59:28 +0000   
531   2016-12-12 00:29:28 +0000   
565   2016-11-25 21:37:47 +0000   
575   2016-11-22 17:28:25 +0000   
705   2016-10-11 00:34:48 +0000   
733   2016-09-29 01:42:20 +0000   
778   2016-09-14 03:27:11 +0000   
822   2016-08-29 03:00:36 +0000   
889   2016-07-31 16:50:42 +0000   
956   2016-07-09 01:08:47 +0000   
1063  2016-06-10 00:39:48 +0000   
1113  2016-05-19 01:38:16 +0000   

                                                 source  \
460   <a href="http://twitter.com/download/iphone" r...   
531   <a href="http://twitter.com/download/iphone" r...   
565   <a href="http://twitter.com/download/iphone" r...   
575   <a href="http://twitter.com/download/iphone" r...   
705   <a href="http://twitter.com/download/iphone" r...   
733   <a href="http://vine.co" rel="nofollow">Vine -...   
778   <a href="http://twitter.com/download/iphone" r...   
822   <a href="http://twitter.com/download/iphone" r...   
889   <a href="http://twitter.com/download/iphone" r...   
956   <a href="http://twitter.com/download/iphone" r...   
1063  <a href="http://twitter.com/download/iphone" r...   
1113  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
460   This is Dido. She's playing the lead role in "...                  NaN   
531   Here we have Burke (pupper) and Dexter (doggo)...                  NaN   
565   Like doggo, like pupper version 2. Both 11/10 ...                  NaN   
575   This is Bones. He's being haunted by another d...                  NaN   
705   This is Pinot. He's a sophisticated doggo. You...                  NaN   
733   Pupper butt 1, Doggo 0. Both 12/10 https://t.c...                  NaN   
778   RT @dog_rates: Like father (doggo), like son (...         7.331095e+17   
822   RT @dog_rates: This is just downright precious...         7.410673e+17   
889   Meet Maggie &amp; Lila. Maggie is the doggo, L...                  NaN   
956   Please stop sending it pictures that don't eve...                  NaN   
1063  This is just downright precious af. 12/10 for ...                  NaN   
1113  Like father (doggo), like son (pupper). Both 1...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
460                        NaN                        NaN   
531                        NaN                        NaN   
565                        NaN                        NaN   
575                        NaN                        NaN   
705                        NaN                        NaN   
733                        NaN                        NaN   
778               4.196984e+09  2016-05-19 01:38:16 +0000   
822               4.196984e+09  2016-06-10 00:39:48 +0000   
889                        NaN                        NaN   
956                        NaN                        NaN   
1063                       NaN                        NaN   
1113                       NaN                        NaN   

                    

Na propria coluna 'text' da descrição acima podemos ver a classificação correta, exceto na classificação 'doggopupper'. Eles são quase todos filhoes com os pais. Vou renomeá-la e alterar as demais.

In [56]:

twitter_links_copy .loc[twitter_links_copy['dog_classification'] == 'doggopuppo', 'dog_classification'] = 'puppo'
twitter_links_copy.loc[twitter_links_copy['dog_classification'] == 'doggofloofer','dog_classification']  = 'floofer'
twitter_links_copy.loc[twitter_links_copy['dog_classification'] == 'doggopupper','dog_classification']  = 'doggo e pupper'

In [57]:
twitter_links_copy['dog_classification'].value_counts()

                  1976
pupper             245
doggo               83
puppo               30
doggo e pupper      12
floofer             10
Name: dog_classification, dtype: int64

Vamos ver a contagem de reetwets

In [58]:
twitter_links_copy.retweeted_status_id.sort_values()

605     6.661041e+17
1043    6.671383e+17
273     6.671522e+17
604     6.671828e+17
603     6.675094e+17
2260    6.675484e+17
2259    6.675487e+17
1023    6.678667e+17
74      6.690004e+17
895     6.703191e+17
602     6.704450e+17
601     6.718968e+17
860     6.732953e+17
811     6.733201e+17
720     6.742918e+17
949     6.753544e+17
600     6.755011e+17
599     6.769365e+17
908     6.790626e+17
926     6.791584e+17
749     6.794628e+17
770     6.798284e+17
506     6.800555e+17
937     6.815232e+17
583     6.816941e+17
655     6.820881e+17
598     6.833919e+17
943     6.835159e+17
654     6.848310e+17
868     6.853251e+17
            ...     
2326             NaN
2327             NaN
2328             NaN
2329             NaN
2330             NaN
2331             NaN
2332             NaN
2333             NaN
2334             NaN
2335             NaN
2336             NaN
2337             NaN
2338             NaN
2339             NaN
2340             NaN
2341             NaN
2342         

In [59]:
twitter_links_copy=twitter_links_copy[twitter_links_copy.tweet_id.notnull()]
twitter_links_copy=twitter_links_copy[twitter_links_copy.retweeted_status_id.isnull()]

In [60]:
twitter_links_copy.retweeted_status_id.nunique()

0

In [61]:
twitter_links_copy.retweeted_status_id.sort_values()

0      NaN
1      NaN
2      NaN
3      NaN
4      NaN
5      NaN
6      NaN
7      NaN
8      NaN
9      NaN
10     NaN
11     NaN
12     NaN
13     NaN
14     NaN
15     NaN
16     NaN
17     NaN
18     NaN
20     NaN
21     NaN
22     NaN
23     NaN
24     NaN
25     NaN
26     NaN
27     NaN
28     NaN
29     NaN
30     NaN
        ..
2326   NaN
2327   NaN
2328   NaN
2329   NaN
2330   NaN
2331   NaN
2332   NaN
2333   NaN
2334   NaN
2335   NaN
2336   NaN
2337   NaN
2338   NaN
2339   NaN
2340   NaN
2341   NaN
2342   NaN
2343   NaN
2344   NaN
2345   NaN
2346   NaN
2347   NaN
2348   NaN
2349   NaN
2350   NaN
2351   NaN
2352   NaN
2353   NaN
2354   NaN
2355   NaN
Name: retweeted_status_id, Length: 2175, dtype: float64

In [62]:
twitter_links_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 14 columns):
tweet_id                      2175 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2175 non-null object
source                        2175 non-null object
text                          2175 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2175 non-null int64
rating_denominator            2175 non-null int64
name                          1495 non-null object
dog_classification            2175 non-null object
dtypes: float64(4), int64(3), object(7)
memory usage: 254.9+ KB


2 - Juntar todas as planilhas em uma planilhas mestra

In [63]:

twitter_to_add_image  = pd.merge(left=twitter_links_copy, right=json_tweets_copy, left_on='tweet_id', right_on='tweet_id', how='inner')


In [64]:

twitter_to_clean  = pd.merge(left=twitter_to_add_image, right=images_copy, left_on='tweet_id', right_on='tweet_id', how='inner')

In [65]:
twitter_to_clean

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888078434458587136                    NaN                  NaN   
20    887705289381826560                    NaN                  NaN   
21    887517139158093824                    NaN                  NaN   
22    887473957103951883                    NaN                  NaN   
23    887343217045368832                    NaN                  NaN   
24    887101392804085760                    NaN                  NaN   
25    886983233522544640                    NaN                  NaN   
26    886736880519319552                    NaN                  NaN   
27    886680336477933568                    NaN                  NaN   
28    886366144734445568                    NaN                  NaN   
29    886258384151887873                    NaN                  NaN   
...                  ...                    ...                  ...   
1993  666020888022790149                    NaN                  NaN   
1994  666020888022790149                    NaN                  NaN   
1995  666020888022790149                    NaN                  NaN   
1996  666020888022790149                    NaN                  NaN   
1997  666020888022790149                    NaN                  NaN   
1998  666020888022790149                    NaN                  NaN   
1999  666020888022790149                    NaN                  NaN   
2000  666020888022790149                    NaN                  NaN   
2001  666020888022790149                    NaN                  NaN   
2002  666020888022790149                    NaN                  NaN   
2003  666020888022790149                    NaN                  NaN   
2004  666020888022790149                    NaN                  NaN   
2005  666020888022790149                    NaN                  NaN   
2006  666020888022790149                    NaN                  NaN   
2007  666020888022790149                    NaN                  NaN   
2008  666020888022790149                    NaN                  NaN   
2009  666020888022790149                    NaN                  NaN   
2010  666020888022790149                    NaN                  NaN   
2011  666020888022790149                    NaN                  NaN   
2012  666020888022790149                    NaN                  NaN   
2013  666020888022790149                    NaN                  NaN   
2014  666020888022790149                    NaN                  NaN   
2015  666020888022790149                    NaN                  NaN   
2016  666020888022790149                    NaN         

Excelente, agora vamos retirar as colunas 'retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp','in_reply_to_status_id','in_reply_to_user_id', 'retweeted_status_id'.

In [169]:
twitter_to_clean = twitter_to_clean.drop(['retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp','in_reply_to_status_id','in_reply_to_user_id', 'retweeted_status_id'],axis=1)

 Vamos alterar os nomes das coluas referentes a raça do animal.

In [68]:
twitter_to_clean.p1 = twitter_to_clean.p1.astype(str)
twitter_to_clean.p2 = twitter_to_clean.p2.astype(str)
twitter_to_clean.p3 = twitter_to_clean.p3.astype(str)

In [69]:
twitter_to_clean.rename(index=str, columns={"p1" :"Previsão_Nº1", "p1_conf":"Confiabilidade","p1_dog" : "Raça",
                                                "p2" :"Previsão_2", "p2_conf":"Confiabilidade","p2_dog" : "Raça",
                                                 "p3" :"Previsão_3", "p3_conf":"Confiabilidade","p3_dog" : "Raça"}, inplace =True)

In [70]:
twitter_to_clean

tweet_id                  timestamp  \
0     892420643555336193  2017-08-01 16:23:56 +0000   
1     892177421306343426  2017-08-01 00:17:27 +0000   
2     891815181378084864  2017-07-31 00:18:03 +0000   
3     891689557279858688  2017-07-30 15:58:51 +0000   
4     891327558926688256  2017-07-29 16:00:24 +0000   
5     891087950875897856  2017-07-29 00:08:17 +0000   
6     890971913173991426  2017-07-28 16:27:12 +0000   
7     890729181411237888  2017-07-28 00:22:40 +0000   
8     890609185150312448  2017-07-27 16:25:51 +0000   
9     890240255349198849  2017-07-26 15:59:51 +0000   
10    890006608113172480  2017-07-26 00:31:25 +0000   
11    889880896479866881  2017-07-25 16:11:53 +0000   
12    889665388333682689  2017-07-25 01:55:32 +0000   
13    889638837579907072  2017-07-25 00:10:02 +0000   
14    889531135344209921  2017-07-24 17:02:04 +0000   
15    889278841981685760  2017-07-24 00:19:32 +0000   
16    888917238123831296  2017-07-23 00:22:39 +0000   
17    888804989199671297  2017-07-22 16:56:37 +0000   
18    888554962724278272  2017-07-22 00:23:06 +0000   
19    888078434458587136  2017-07-20 16:49:33 +0000   
20    887705289381826560  2017-07-19 16:06:48 +0000   
21    887517139158093824  2017-07-19 03:39:09 +0000   
22    887473957103951883  2017-07-19 00:47:34 +0000   
23    887343217045368832  2017-07-18 16:08:03 +0000   
24    887101392804085760  2017-07-18 00:07:08 +0000   
25    886983233522544640  2017-07-17 16:17:36 +0000   
26    886736880519319552  2017-07-16 23:58:41 +0000   
27    886680336477933568  2017-07-16 20:14:00 +0000   
28    886366144734445568  2017-07-15 23:25:31 +0000   
29    886258384151887873  2017-07-15 16:17:19 +0000   
...                  ...                        ...   
1993  666020888022790149  2015-11-15 22:32:08 +0000   
1994  666020888022790149  2015-11-15 22:32:08 +0000   
1995  666020888022790149  2015-11-15 22:32:08 +0000   
1996  666020888022790149  2015-11-15 22:32:08 +0000   
1997  666020888022790149  2015-11-15 22:32:08 +0000   
1998  666020888022790149  2015-11-15 22:32:08 +0000   
1999  666020888022790149  2015-11-15 22:32:08 +0000   
2000  666020888022790149  2015-11-15 22:32:08 +0000   
2001  666020888022790149  2015-11-15 22:32:08 +0000   
2002  666020888022790149  2015-11-15 22:32:08 +0000   
2003  666020888022790149  2015-11-15 22:32:08 +0000   
2004  666020888022790149  2015-11-15 22:32:08 +0000   
2005  666020888022790149  2015-11-15 22:32:08 +0000   
2006  666020888022790149  2015-11-15 22:32:08 +0000   
2007  666020888022790149  2015-11-15 22:32:08 +0000   
2008  666020888022790149  2015-11-15 22:32:08 +0000   
2009  666020888022790149  2015-11-15 22:32:08 +0000   
2010  666020888022790149  2015-11-15 22:32:08 +0000   
2011  666020888022790149  2015-11-15 22:32:08 +0000   
2012  666020888022790149  2015-11-15 22:32:08 +0000   
2013  666020888022790149  2015-11-15 22:32:08 +0000   
2014  666020888022790149  2015-11-15 22:32:08 +0000   
2015  666020888022790149  2015-11-15 22:32:08 +0000   
2016  666020888022790149  2015-11-15 22:32:08 +0000   
2017  666020888022790149  2015-11-15 22:32:08 +0000   
2018  666020888022790149  2015-11-15 22:32:08 +0000   
2019  666020888022790149  2015-11-15 22:32:08 +0000   
2020  666020888022790149  2015-11-15 22:32:08 +0000   
2021  666020888022790149  2015-11-15 22:32:08 +0000   
2022  666020888022790149  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
5     <a href="http://twitter.com/download/iphone" r...   
6     <a href="http://twitter.com/download/iphone" r...   
7     <a href="http://twitter.com/download/iphone" r...   
8     <a href="http://twitter.com/download/iphone" r...   
9     <a href="

Vamos tratar a coluna nome, que tem varias palavras e letras sem sentido, e classificá-las como Nan.


In [75]:
twitter_to_clean.name = twitter_to_clean.name.replace(['a','an','None','the', 'quite', 'not', 'officially', 'space', 'such', 'unacceptable', 'this', 'very'], np.nan)

In [76]:
twitter_to_clean.name.sort_values()

770             Abby
759             Abby
692              Ace
1578            Acro
1031           Adele
1579           Aiden
65               Aja
354            Akumi
597               Al
641           Albert
1598          Albert
116            Albus
308            Albus
842          Aldrick
1689       Alejandro
280        Alexander
914     Alexanderson
359              Alf
1286           Alfie
273            Alfie
627            Alfie
1878           Alfie
45              Alfy
155            Alice
1363           Alice
1181           Amber
1038         Ambrose
1721             Amy
1836          Amélie
565           Anakin
            ...     
1993             NaN
1994             NaN
1995             NaN
1996             NaN
1997             NaN
1998             NaN
1999             NaN
2000             NaN
2001             NaN
2002             NaN
2003             NaN
2004             NaN
2005             NaN
2006             NaN
2007             NaN
2008             NaN
2009         

In [77]:
twitter_to_clean.tweet_id.count()

2023

In [78]:
twitter_to_clean[twitter_to_clean.duplicated(['jpg_url'], keep=False)]['jpg_url']

1992    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg
1993    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg
1994    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg
1995    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg
1996    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg
1997    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg
1998    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg
1999    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg
2000    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg
2001    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg
2002    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg
2003    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg
2004    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg
2005    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg
2006    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg
2007    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg
2008    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg
2009    https://pbs.twimg.com/media/CT4udn0WwAA0

In [79]:
twitter_to_clean.expanded_urls.count()

2023

In [82]:
twitter_to_clean.expanded_urls.sort_values()

978     http://goo.gl/ArWZfi,https://twitter.com/dog_r...
293     http://us.blastingnews.com/news/2017/01/kentuc...
258     http://www.gofundme.com/bluethewhitehusky,http...
6       https://gofundme.com/ydvmve-surgery-for-jax,ht...
378     https://m.facebook.com/story.php?story_fbid=18...
2001    https://twitter.com/dog_rates/status/666020888...
2000    https://twitter.com/dog_rates/status/666020888...
1999    https://twitter.com/dog_rates/status/666020888...
1998    https://twitter.com/dog_rates/status/666020888...
1997    https://twitter.com/dog_rates/status/666020888...
1994    https://twitter.com/dog_rates/status/666020888...
1995    https://twitter.com/dog_rates/status/666020888...
2002    https://twitter.com/dog_rates/status/666020888...
1993    https://twitter.com/dog_rates/status/666020888...
1992    https://twitter.com/dog_rates/status/666020888...
2021    https://twitter.com/dog_rates/status/666020888...
1996    https://twitter.com/dog_rates/status/666020888...
2003    https:

In [81]:
twitter_to_clean.nunique()

tweet_id              1993
timestamp             1993
source                   3
text                  1993
expanded_urls         1993
rating_numerator        34
rating_denominator      15
name                   924
dog_classification       6
curtidas              1866
retweets              1593
data                  1993
jpg_url               1993
img_num                  4
Previsão_Nº1           374
Confiabilidade        1990
Raça                     2
Previsão_2             401
Confiabilidade        1988
Raça                     2
Previsão_3             405
Confiabilidade        1990
Raça                     2
dtype: int64

In [254]:
twitter_to_clean.tweet_id.sort_values()

2022    666020888022790149
1992    666020888022790149
1993    666020888022790149
1994    666020888022790149
1995    666020888022790149
1996    666020888022790149
1997    666020888022790149
1998    666020888022790149
1999    666020888022790149
2000    666020888022790149
2001    666020888022790149
2002    666020888022790149
2003    666020888022790149
2004    666020888022790149
2005    666020888022790149
2021    666020888022790149
2007    666020888022790149
2020    666020888022790149
2019    666020888022790149
2018    666020888022790149
2017    666020888022790149
2016    666020888022790149
2015    666020888022790149
2006    666020888022790149
2013    666020888022790149
2008    666020888022790149
2009    666020888022790149
2014    666020888022790149
2011    666020888022790149
2010    666020888022790149
               ...        
29      886258384151887873
28      886366144734445568
27      886680336477933568
26      886736880519319552
25      886983233522544640
24      887101392804085760
2

Novo problema de limpeza - ainda temos Id's repetidos no Dataset provenientes do arquivo imagens.
Vamos limpá-los.

In [83]:
twitter_to_clean = twitter_to_clean.drop_duplicates(['tweet_id'])


In [84]:
twitter_to_clean.nunique()

tweet_id              1993
timestamp             1993
source                   3
text                  1993
expanded_urls         1993
rating_numerator        34
rating_denominator      15
name                   924
dog_classification       6
curtidas              1866
retweets              1593
data                  1993
jpg_url               1993
img_num                  4
Previsão_Nº1           374
Confiabilidade        1990
Raça                     2
Previsão_2             401
Confiabilidade        1988
Raça                     2
Previsão_3             405
Confiabilidade        1990
Raça                     2
dtype: int64

In [85]:
twitter_to_clean.jpg_url.nunique()

1993

In [86]:
twitter_to_clean = twitter_to_clean.drop_duplicates(['jpg_url'])

In [87]:
twitter_to_clean.nunique()

tweet_id              1993
timestamp             1993
source                   3
text                  1993
expanded_urls         1993
rating_numerator        34
rating_denominator      15
name                   924
dog_classification       6
curtidas              1866
retweets              1593
data                  1993
jpg_url               1993
img_num                  4
Previsão_Nº1           374
Confiabilidade        1990
Raça                     2
Previsão_2             401
Confiabilidade        1988
Raça                     2
Previsão_3             405
Confiabilidade        1990
Raça                     2
dtype: int64

Otimo. Agora temos apenas Tweets originais e com imagens.

In [90]:
twitter_to_clean.rating_numerator.sort_values()

235        0
755        0
1734       1
1518       1
1899       1
1972       1
1975       1
1722       2
1565       2
1421       2
1877       2
1886       2
1947       2
1963       2
1418       2
1986       2
1879       3
1486       3
1825       3
1573       3
1583       3
1166       3
1591       3
1942       3
1843       3
1779       3
554        3
1105       3
1354       3
964        3
        ... 
549       14
131       14
571       14
49        14
399       14
147       14
146       14
362       14
253       14
64        14
256       14
385       24
1373      26
553       27
1129      44
987       45
924       50
1053      60
503       75
969       80
323       84
1493      88
945       99
1302     121
1301     143
1434     144
662      165
847      204
1717     420
725     1776
Name: rating_numerator, Length: 1993, dtype: int64

In [91]:
twitter_to_clean.rating_denominator.sort_values()

1972      2
385       7
0        10
1336     10
1335     10
1334     10
1333     10
1332     10
1331     10
1330     10
1329     10
1328     10
1326     10
1325     10
1324     10
1323     10
1322     10
1321     10
1320     10
1319     10
1318     10
1317     10
1316     10
1315     10
1314     10
1313     10
1312     10
1311     10
1310     10
1337     10
       ... 
671      10
656      10
669      10
668      10
667      10
666      10
670      10
665      10
664      10
658      10
663      10
659      10
660      10
661      10
1327     11
799      11
890      20
1129     40
924      50
987      50
1053     50
323      70
969      80
1493     80
945      90
1302    110
1434    120
1301    130
662     150
847     170
Name: rating_denominator, Length: 1993, dtype: int64

Para numeradores e denominadores vamos seguir o seguinte processo:
- Os denominadores geralmente são com o número 10. Vamos ver denominadores muito fora do padrão.
- Os numeradores não tem base definida, vamos checar os que estão fora do padrão
- Vamos checar nos textos por valores decimais.
- Mudar os valores para float

In [118]:
print(twitter_to_clean[twitter_to_clean.rating_denominator == 2]['text'])
print(twitter_to_clean[twitter_to_clean.rating_denominator == 7]['text'])
print(twitter_to_clean[twitter_to_clean.rating_denominator == 11]['text'])
print(twitter_to_clean[twitter_to_clean.rating_denominator == 20]['text'])
print(twitter_to_clean[twitter_to_clean.rating_denominator == 40]['text'])
print(twitter_to_clean[twitter_to_clean.rating_denominator == 50]['text'])
print(twitter_to_clean[twitter_to_clean.rating_denominator == 110]['text'])
print(twitter_to_clean[twitter_to_clean.rating_denominator == 120]['text'])
print(twitter_to_clean[twitter_to_clean.rating_denominator == 130]['text'])
print(twitter_to_clean[twitter_to_clean.rating_denominator == 150]['text'])
print(twitter_to_clean[twitter_to_clean.rating_denominator == 170]['text'])
print(twitter_to_clean[twitter_to_clean.rating_numerator == 1776]['text'])
print(twitter_to_clean[twitter_to_clean.rating_numerator == 420]['text'])
print(twitter_to_clean[twitter_to_clean.rating_numerator == 204]['text'])


1972    This is an Albanian 3 1/2 legged  Episcopalian...
Name: text, dtype: object
385    Meet Sam. She smiles 24/7 &amp; secretly aspir...
Name: text, dtype: object
799     After so many requests, this is Bretagne. She ...
1327    This is Darrel. He just robbed a 7/11 and is i...
Name: text, dtype: object
890    Happy 4/20 from the squad! 13/10 for all https...
Name: text, dtype: object
1129    Happy Wednesday here's a bucket of pups. 44/40...
Name: text, dtype: object
924     This is Bluebert. He just saw that both #Final...
987     From left to right:\nCletus, Jerome, Alejandro...
1053    Here is a whole flock of puppers.  60/50 I'll ...
Name: text, dtype: object
1302    Someone help the girl is being mugged. Several...
Name: text, dtype: object
1434    IT'S PUPPERGEDDON. Total of 144/120 ...I think...
Name: text, dtype: object
1301    Two sneaky puppers were not initially seen, mo...
Name: text, dtype: object
662    Why does this never happen at my front door......
Name: text, dty

In [157]:
twitter_to_clean.loc[twitter_to_clean['rating_denominator'] == 2,['rating_numerator','rating_denominator']] = 9,10
twitter_to_clean.loc[twitter_to_clean['rating_denominator'] == 11,['rating_numerator','rating_denominator']] = 14,10
twitter_to_clean.loc[twitter_to_clean['rating_denominator'] == 20,['rating_numerator','rating_denominator']] = 13,10
twitter_to_clean.loc[twitter_to_clean['rating_numerator'] == 50,['rating_numerator','rating_denominator']] = 11,10



666287406224695296 - Alterado. Os valores 1/2 estavam na legenda mas não eram a nota.

810984652412424192 - Está ok

740373189193256964 - Alterado.

722974582966214656 - Alterado.

697463031882764288 - ok.

709198395643068416 - ok.

704054845121142784 - ok.

716439118184652801 - Alterado.

684222868335505415 - ok.

677716515794329600 - ok.

758467244762497024 - ok.

731156023742988288 - ok.

749981277374128128 - ok.

670842764863651840 - ok.

In [158]:
#Vamos checar erros atraves dos textos
twitter_to_clean[twitter_to_clean.text.str.contains(r"(\d+\.\d*\/\d+\.?)")]

C:\Users\Samsung\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  


tweet_id                  timestamp  \
39    883482846933004288  2017-07-08 00:28:19 +0000   
503   786709082849828864  2016-10-13 23:23:56 +0000   
553   778027034220126208  2016-09-20 00:24:34 +0000   
1373  680494726643068929  2015-12-25 21:06:00 +0000   

                                                 source  \
39    <a href="http://twitter.com/download/iphone" r...   
503   <a href="http://twitter.com/download/iphone" r...   
553   <a href="http://twitter.com/download/iphone" r...   
1373  <a href="http://twitter.com/download/iphone" r...   

                                                   text  \
39    This is Bella. She hopes her smile made you sm...   
503   This is Logan, the Chow who lived. He solemnly...   
553   This is Sophie. She's a Jubilant Bush Pupper. ...   
1373  Here we have uncovered an entire battalion of ...   

                                          expanded_urls  rating_numerator  \
39    https://twitter.com/dog_rates/status/883482846...               5.0   
503   https://twitter.com/dog_rates/status/786709082...              75.0   
553   https://twitter.com/dog_rates/status/778027034...              27.0   
1373  https://twitter.com/dog_rates/status/680494726...              26.0   

      rating_denominator    name dog_classification  curtidas  ...   img_num  \
39                  10.0   Bella                        45493  ...         1   
503                 10.0   Logan                        19634  ...         1   
553                 10.0  Sophie             pupper      7059  ...         1   
1373                10.0     NaN                         1809  ...         1   

          Previsão_Nº1  Confiabilidade  Raça          Previsão_2  \
39    golden_retriever        0.943082  True  Labrador_retriever   
503         Pomeranian        0.467321  True         Persian_cat   
553            clumber        0.946718  True      cocker_spaniel   
1373            kuvasz        0.438627  True             Samoyed   

      Confiabilidade   Raça      Previsão_3  Confiabilidade  Raça  
39          0.032409   True          kuvasz        0.005501  True  
503         0.122978  False            chow        0.102654  True  
553         0.015950   True           Lhasa        0.006519  True  
1373        0.111622   True  Great_Pyrenees        0.064061  True  

[4 rows x 23 columns]

In [159]:
#Definir os campos como float
twitter_to_clean['rating_numerator'] = twitter_to_clean['rating_numerator'].astype('float')
twitter_to_clean['rating_denominator'] = twitter_to_clean['rating_denominator'].astype('float')

Vamos criar uma coluna de 'Avaliação', que fará o calculo do numerador pelo denominador. Assim poderemos usar estes valores para análise mais tarde.

In [160]:
twitter_to_clean['Avaliação'] = twitter_to_clean['rating_numerator']/twitter_to_clean['rating_denominator']

In [162]:
twitter_to_clean

tweet_id                  timestamp  \
0     892420643555336193  2017-08-01 16:23:56 +0000   
1     892177421306343426  2017-08-01 00:17:27 +0000   
2     891815181378084864  2017-07-31 00:18:03 +0000   
3     891689557279858688  2017-07-30 15:58:51 +0000   
4     891327558926688256  2017-07-29 16:00:24 +0000   
5     891087950875897856  2017-07-29 00:08:17 +0000   
6     890971913173991426  2017-07-28 16:27:12 +0000   
7     890729181411237888  2017-07-28 00:22:40 +0000   
8     890609185150312448  2017-07-27 16:25:51 +0000   
9     890240255349198849  2017-07-26 15:59:51 +0000   
10    890006608113172480  2017-07-26 00:31:25 +0000   
11    889880896479866881  2017-07-25 16:11:53 +0000   
12    889665388333682689  2017-07-25 01:55:32 +0000   
13    889638837579907072  2017-07-25 00:10:02 +0000   
14    889531135344209921  2017-07-24 17:02:04 +0000   
15    889278841981685760  2017-07-24 00:19:32 +0000   
16    888917238123831296  2017-07-23 00:22:39 +0000   
17    888804989199671297  2017-07-22 16:56:37 +0000   
18    888554962724278272  2017-07-22 00:23:06 +0000   
19    888078434458587136  2017-07-20 16:49:33 +0000   
20    887705289381826560  2017-07-19 16:06:48 +0000   
21    887517139158093824  2017-07-19 03:39:09 +0000   
22    887473957103951883  2017-07-19 00:47:34 +0000   
23    887343217045368832  2017-07-18 16:08:03 +0000   
24    887101392804085760  2017-07-18 00:07:08 +0000   
25    886983233522544640  2017-07-17 16:17:36 +0000   
26    886736880519319552  2017-07-16 23:58:41 +0000   
27    886680336477933568  2017-07-16 20:14:00 +0000   
28    886366144734445568  2017-07-15 23:25:31 +0000   
29    886258384151887873  2017-07-15 16:17:19 +0000   
...                  ...                        ...   
1963  666411507551481857  2015-11-17 00:24:19 +0000   
1964  666407126856765440  2015-11-17 00:06:54 +0000   
1965  666396247373291520  2015-11-16 23:23:41 +0000   
1966  666373753744588802  2015-11-16 21:54:18 +0000   
1967  666362758909284353  2015-11-16 21:10:36 +0000   
1968  666353288456101888  2015-11-16 20:32:58 +0000   
1969  666345417576210432  2015-11-16 20:01:42 +0000   
1970  666337882303524864  2015-11-16 19:31:45 +0000   
1971  666293911632134144  2015-11-16 16:37:02 +0000   
1972  666287406224695296  2015-11-16 16:11:11 +0000   
1973  666273097616637952  2015-11-16 15:14:19 +0000   
1974  666268910803644416  2015-11-16 14:57:41 +0000   
1975  666104133288665088  2015-11-16 04:02:55 +0000   
1976  666102155909144576  2015-11-16 03:55:04 +0000   
1977  666099513787052032  2015-11-16 03:44:34 +0000   
1978  666094000022159362  2015-11-16 03:22:39 +0000   
1979  666082916733198337  2015-11-16 02:38:37 +0000   
1980  666073100786774016  2015-11-16 01:59:36 +0000   
1981  666071193221509120  2015-11-16 01:52:02 +0000   
1982  666063827256086533  2015-11-16 01:22:45 +0000   
1983  666058600524156928  2015-11-16 01:01:59 +0000   
1984  666057090499244032  2015-11-16 00:55:59 +0000   
1985  666055525042405380  2015-11-16 00:49:46 +0000   
1986  666051853826850816  2015-11-16 00:35:11 +0000   
1987  666050758794694657  2015-11-16 00:30:50 +0000   
1988  666049248165822465  2015-11-16 00:24:50 +0000   
1989  666044226329800704  2015-11-16 00:04:52 +0000   
1990  666033412701032449  2015-11-15 23:21:54 +0000   
1991  666029285002620928  2015-11-15 23:05:30 +0000   
1992  666020888022790149  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
5     <a href="http://twitter.com/download/iphone" r...   
6     <a href="http://twitter.com/download/iphone" r...   
7     <a href="http://twitter.com/download/iphone" r...   
8     <a href="http://twitter.com/download/iphone" r...   
9     <a href="

Excelente.

In [163]:
twitter_to_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1993 entries, 0 to 1992
Data columns (total 24 columns):
tweet_id              1993 non-null object
timestamp             1993 non-null object
source                1993 non-null object
text                  1993 non-null object
expanded_urls         1993 non-null object
rating_numerator      1993 non-null float64
rating_denominator    1993 non-null float64
name                  1366 non-null object
dog_classification    1993 non-null object
curtidas              1993 non-null int64
retweets              1993 non-null int64
data                  1993 non-null object
jpg_url               1993 non-null object
img_num               1993 non-null int64
Previsão_Nº1          1993 non-null object
Confiabilidade        1993 non-null float64
Raça                  1993 non-null bool
Previsão_2            1993 non-null object
Confiabilidade        1993 non-null float64
Raça                  1993 non-null bool
Previsão_3            1993 non-null objec

In [164]:
twitter_to_clean.describe()

rating_numerator  rating_denominator       curtidas      retweets  \
count       1993.000000         1993.000000    1993.000000   1993.000000   
mean          12.275966           10.510286    8793.519318   2683.528851   
std           41.498223            7.263336   12814.528242   4758.689626   
min            0.000000            7.000000      80.000000     12.000000   
25%           10.000000           10.000000    1897.000000    597.000000   
50%           11.000000           10.000000    3990.000000   1293.000000   
75%           12.000000           10.000000   10996.000000   3081.000000   
max         1776.000000          170.000000  164929.000000  84508.000000   

           img_num  Confiabilidade  Confiabilidade  Confiabilidade  \
count  1993.000000     1993.000000    1.993000e+03    1.993000e+03   
mean      1.203211        0.593937    1.343498e-01    6.025022e-02   
std       0.560899        0.272022    1.006579e-01    5.090338e-02   
min       1.000000        0.044333    1.011300e-08    1.740170e-10   
25%       1.000000        0.362835    5.390140e-02    1.619070e-02   
50%       1.000000        0.587507    1.174020e-01    4.948690e-02   
75%       1.000000        0.846628    1.948970e-01    9.160200e-02   
max       4.000000        1.000000    4.880140e-01    2.734190e-01   

         Avaliação  
count  1993.000000  
mean      1.170712  
std       4.067991  
min       0.000000  
25%       1.000000  
50%       1.100000  
75%       1.200000  
max     177.600000

No final temos 1993 registros para análise.

Vamos alterar o tipo de dado de  timestamp(), para que possamos trabalhar com essa variavel depois, se possivel.


In [165]:

twitter_to_clean['timestamp'] = pd.to_datetime(twitter_to_clean['timestamp'])



In [168]:
twitter_to_clean.timestamp.sort_values()

1992   2015-11-15 22:32:08
1991   2015-11-15 23:05:30
1990   2015-11-15 23:21:54
1989   2015-11-16 00:04:52
1988   2015-11-16 00:24:50
1987   2015-11-16 00:30:50
1986   2015-11-16 00:35:11
1985   2015-11-16 00:49:46
1984   2015-11-16 00:55:59
1983   2015-11-16 01:01:59
1982   2015-11-16 01:22:45
1981   2015-11-16 01:52:02
1980   2015-11-16 01:59:36
1979   2015-11-16 02:38:37
1978   2015-11-16 03:22:39
1977   2015-11-16 03:44:34
1976   2015-11-16 03:55:04
1975   2015-11-16 04:02:55
1974   2015-11-16 14:57:41
1973   2015-11-16 15:14:19
1972   2015-11-16 16:11:11
1971   2015-11-16 16:37:02
1970   2015-11-16 19:31:45
1969   2015-11-16 20:01:42
1968   2015-11-16 20:32:58
1967   2015-11-16 21:10:36
1966   2015-11-16 21:54:18
1965   2015-11-16 23:23:41
1964   2015-11-17 00:06:54
1963   2015-11-17 00:24:19
               ...        
29     2017-07-15 16:17:19
28     2017-07-15 23:25:31
27     2017-07-16 20:14:00
26     2017-07-16 23:58:41
25     2017-07-17 16:17:36
24     2017-07-18 00:07:08
2

Vamos analisar os dados.

In [167]:

twitter_to_clean.to_csv('twitter_archive_master.csv')